### PR TITLE
Transactions support

### PR DESCRIPTION
**Pull request by ziaratban**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes

### Examples
```php
#Starts and commits transaction in easy mode 
yii::$app->mongodb->transaction(function($clientSession){
    #Your codes in transaction
},[
    'readConcern' => 'snapshot',
    'writeConcern' => 'majority',
]);
```
```php
#Supports nested calling (this is not nested transaction!)
yii::$app->mongodb->transaction(function(){

    User::deleteAll(...);

    yii::$app->mongodb->transaction(function(){
        #Execute your mongodb commands in new session and transaction
        User::deleteAll(...);

        yii::$app->mongodb->noTransaction(function(){
             #Execute your mongodb commands in out of session and transaction
             //your codes ...
        });
    }); 

    return false; #Roll backing the current transaction
});
```
```php
#Other way
yii::$app->mongodb->startTransaction([
    'readConcern' => 'snapshot', #transaction options
],[
    'causalConsistency' => true, #session options
]);

#your codes ...

yii::$app->mongodb->commitTransaction(); #or yii::$app->mongodb->rollBackTransaction();
```
**`Once` feature**
Starts and commits transaction in easy mode if the previous transaction was not executed, otherwise only runs your actions in previous transaction.
```php
yii::$app->mongodb->transaction(function(){
    yii::$app->mongodb->transactionOnce(function(){
        #Execute your mongodb commands in previous transaction.
    });
});
```
```php
yii::$app->mongodb->startSession(); #Ends the previous session and starts the new session.
yii::$app->mongodb->startSessionOnce(); #The session does not start.
```
[**Lock a document for update**](https://www.mongodb.com/blog/post/how-to-select--for-update-inside-mongodb-transactions)
```php
yii::$app->mongodb->transaction(function($clientSession){
    #Locks a document of the collection in a transaction(like `select for update` feature in mysql)
    try{
        $doc = MyActiveRecordClass::LockDocument(new ObjectId('5e5fa6125845350caf596ed6'), '_lock');
        if($doc === null)
            echo 'document not found!';
        else{
            #The document is locked and you can have a safe update
        }
    }
    catch(\Exception $e){
        #see https://github.com/yiisoft/yii2-mongodb/pull/306
        if(ErrorCode::is($e,ErrorCode::WriteConflict))
            echo 'This is a mongodb write conflict error.';
    }
},[
    'readConcern' => 'snapshot',
    'writeConcern' => 'majority',
]);
```
**Stubbornly lock feature**
When doc1 is locked in a transaction(t1) and when other transactions(t2) want to update doc1, they will confront 'writeConflict' error. if you don't want to have a  'writeConflict' error , you can use the 'stubbornly lock' feature in application level.
this feature will wait until the document unlocks.
```php
#notice : you can not use stubborn mode if transaction is started in current session(or use your session with `mySession` parameter).
#starts new transaction in current session(create new session if it doesn't exist)
$doc = MyActiveRecordClass::LockDocumentStubbornly('5e652b4a81eb9b37cac99a2b','_lock',[
    'mySession' => [],        #a custom session instance of ClientSession for starting a transaction.
    'modifyOptions' => [],    #see $options in ActiveQuery::modify()
    'sleep' => 1000000,       #a time parameter in microseconds for waiting. the default is one second.
    'try' => 0,               #maximum number of retrying. throw write conflict error after it has reached this value. the zero default is unlimited.
    'transactionOptions' => [ #new transaction options. see $transactionOptions in Transaction::start()
        'readConcern' => 'majority',
        'writeConcern' => 'majority',
    ]
]);
$doc->f1++;
$doc->Save();
yii::$app->mongodb->CommitTransaction();
```

### CODE REVIEW

**Comment by samdark** [comment]
Tests seems to fail: https://github.com/yiisoft/yii2-mongodb/pull/294/checks?check_run_id=484726582

**Comment by ziaratban** [comment]
@samdark for fixing this [errors](https://travis-ci.com/yiisoft/yii2-mongodb/jobs/294055460) , i must be change version of mongodb to 4.0 (also mongodb php driver to 1.5).
do problem solved if i change [.travis.yml:28](https://github.com/yiisoft/yii2-mongodb/blob/master/.travis.yml#L28) to `- mongodb-4.0-precise` ?

please see [MongoDB Compatibility](https://docs.mongodb.com/ecosystem/drivers/php/#mongodb-compatibility)

**Comment by samdark** [comment]
@ziaratban you can try adjusting `.travis.yml` and pushing it. Travis will automatically pick it up so you'll see results after test run.

**Comment by ziaratban** [comment]
@samdark 
Test on php `5.5` and `5.6` fails because the test environment has bugs.
please see #316 .

File: src/ActiveRecord.php [hunk 1]
```diff
@@ -208,8 +230,14 @@ public function insert($runValidation = true, $attributes = null)
...
         }
-        $result = $this->insertInternal($attributes);
 
+        if(!$this->isTransactional(self::OP_INSERT))
+            return $this->insertInternal($attributes);
+
+        $result = null;
+        static::getDb()->transaction(function()use($attribute,&$result){
+            $result = $this->insertInternal($attributes);
+        });
```
Comment (samdark) [inline src/ActiveRecord.php:240]:
```suggestion
        if (!$this->isTransactional(self::OP_INSERT)) {
            return $this->insertInternal($attributes);
        }

        $result = null;
        static::getDb()->transaction(function() use ($attribute, &$result) {
            $result = $this->insertInternal($attributes);
        });
```

**Comment by Jafarili** [comment]
Thank you @ziaratban for the PR
Is there any update for this?

**Comment by ziaratban** [comment]
> Thank you @ziaratban for the PR Is there any update for this?

I think it is over and I am waiting for @samdark to have a code review and I have been using this feature in my products for about a year now.

If you have a code to add, we will be very happy.

**Comment by samdark** [comment]
@ziaratban thanks for patience and sorry it takes so long.

File: src/ActiveRecord.php [hunk 2]
```diff
@@ -411,4 +516,146 @@ private function dumpBsonObject(Type $object)
         }
         return ArrayHelper::toArray($object);
     }
-}
+
+    /**
+     * Locks a document of the collection in a transaction(like `select for update` feature in mysql)
+     * @see https://www.mongodb.com/blog/post/how-to-select--for-update-inside-mongodb-transactions
+     * @param mixed $id a document id(primary key > _id)
```
Comment (samdark) [inline src/ActiveRecord.php:523]:
```suggestion
     * @param mixed $id a document id (primary key > _id)
```

File: src/ActiveRecord.php [hunk 3]
```diff
@@ -411,4 +516,146 @@ private function dumpBsonObject(Type $object)
...
+                $modifyOptions,
+                $db
+            )
+        ;
+    }
+
+    /**
+     * Locking a document in stubborn mode on a transaction (like `select for update` feature in MySQL)
+     * @see https://www.mongodb.com/blog/post/how-to-select--for-update-inside-mongodb-transactions
+     * notice : you can not use stubborn mode if transaction is started in current session(or use your session with `mySession` parameter).
```
Comment (samdark) [inline src/ActiveRecord.php:550]:
```suggestion
     * notice : you can not use stubborn mode if transaction is started in current session (or use your session with `mySession` parameter).
```

File: src/ActiveRecord.php [hunk 4]
```diff
@@ -411,4 +516,146 @@ private function dumpBsonObject(Type $object)
...
+     * @param mixed $id a document id(primary key > _id)
```
Comment (samdark) [inline src/ActiveRecord.php:551]:
```suggestion
     * @param mixed $id a document id (primary key > _id)
```

File: src/ActiveRecord.php [hunk 5]
```diff
@@ -411,4 +516,146 @@ private function dumpBsonObject(Type $object)
...
+     * @param array $options list of options in format:
+     *   [
+     *     'mySession' => false,        #A custom session instance of ClientSession for start a transaction.
+     *     'transactionOptions' => [],  #New transaction options. see $transactionOptions in Transaction::start()
+     *     'modifyOptions' => [],       #See $options in ActiveQuery::modify()
+     *     'sleep' => 1000000,          #A time parameter in microseconds to wait. the default is one second.
+     *     'try' => 0,                  #Maximum count of retry. throw write conflict error after reached this value. the zero default is unlimited.
+     *     'lockFieldName' => '_lock'   #The name of the field you want to lock. default is '_lock'
```
Comment (samdark) [inline src/ActiveRecord.php:559]:
```suggestion
     *     'mySession' => false,        # A custom session instance of ClientSession for start a transaction.
     *     'transactionOptions' => [],  # New transaction options. see $transactionOptions in Transaction::start()
     *     'modifyOptions' => [],       # See $options in ActiveQuery::modify()
     *     'sleep' => 1000000,          # A time parameter in microseconds to wait. the default is one second.
     *     'try' => 0,                  # Maximum count of retry. throw write conflict error after reached this value. the zero default is unlimited.
     *     'lockFieldName' => '_lock'   # The name of the field you want to lock. default is '_lock'
```

File: src/Collection.php [hunk 6]
```diff
@@ -56,23 +56,25 @@ public function getFullName()
 
     /**
      * Drops this collection.
+     * @param array $execOptions -> goto Command::execute()
```
Comment (samdark) [inline src/Collection.php:59]: Need a better description about what could be passed in this parameter.

File: src/Collection.php [hunk 7]
```diff
@@ -56,23 +56,25 @@ public function getFullName()
...
      * @throws Exception on failure.
      * @return bool whether the operation successful.
      */
-    public function drop()
+    public function drop($execOptions = [])
```
Comment (samdark) [inline src/Collection.php:63]: Need a better description about what could be passed in this parameter.

File: src/Collection.php [hunk 8]
```diff
@@ -56,23 +56,25 @@ public function getFullName()
...
     {
-        return $this->database->dropCollection($this->name);
+        return $this->database->dropCollection($this->name, $execOptions);
     }
 
     /**
      * Returns the list of defined indexes.
      * @return array list of indexes info.
      * @param array $options list of options in format: optionName => optionValue.
+     * @param array $execOptions -> goto Command::execute()
```
Comment (samdark) [inline src/Collection.php:72]: Need a better description about what could be passed in this parameter.

File: src/Collection.php [hunk 9]
```diff
@@ -107,23 +109,25 @@ public function listIndexes($options = [])
      *
      * See [[https://docs.mongodb.com/manual/reference/method/db.collection.createIndex/#options-for-all-index-types]]
      * for the full list of options.
+     * @param array $execOptions -> goto Command::execute()
```
Comment (samdark) [inline src/Collection.php:112]: Need a better description about what could be passed in this parameter.

File: src/Collection.php [hunk 10]
```diff
@@ -107,23 +109,25 @@ public function listIndexes($options = [])
...
     {
-        return $this->database->createCommand()->createIndexes($this->name, $indexes);
+        return $this->database->createCommand()->createIndexes($this->name, $indexes, $execOptions);
     }
 
     /**
      * Drops collection indexes by name.
      * @param string $indexes wildcard for name of the indexes to be dropped.
      * You can use `*` to drop all indexes.
+     * @param array $execOptions -> goto Command::execute()
```
Comment (samdark) [inline src/Collection.php:125]: Need a better description about what could be passed in this parameter.

File: src/Collection.php [hunk 11]
```diff
@@ -144,13 +148,14 @@ public function dropIndexes($indexes)
      * ```
      *
      * @param array $options list of options in format: optionName => optionValue.
+     * @param array $execOptions -> goto Command::execute()
```
Comment (samdark) [inline src/Collection.php:151]: Need a better description about what could be passed in this parameter.

File: src/Collection.php [hunk 12]
```diff
@@ -171,17 +176,18 @@ public function createIndex($columns, $options = [])
      * ]
      * ```
      *
+     * @param array $execOptions -> goto Command::execute()
```
Comment (samdark) [inline src/Collection.php:179]: Need a better description about what could be passed in this parameter.

File: src/Collection.php [hunk 13]
```diff
@@ -200,12 +206,13 @@ public function dropIndex($columns)
 
     /**
      * Drops all indexes for this collection.
+     * @param array $execOptions -> goto Command::execute()
```
Comment (samdark) [inline src/Collection.php:209]: Need a better description about what could be passed in this parameter.

File: src/Collection.php [hunk 14]
```diff
@@ -215,28 +222,30 @@ public function dropAllIndexes()
      * @param array $condition query condition
      * @param array $fields fields to be selected
      * @param array $options query options (available since 2.1).
+     * @param array $execOptions -> goto Command::executeQuery()
```
Comment (samdark) [inline src/Collection.php:225]: Need a better description about what could be passed in this parameter. Alternatively, {@see Command::executeQuery()} could be used.

File: src/Command.php [hunk 15]
```diff
@@ -74,106 +74,74 @@ class Command extends BaseObject
...
-    public function setWriteConcern($writeConcern)
+    * preapare Concern and Preference options for easy use
+    * @param array|object by reference
+    * convert string option to object
+    * ['readConcern' => 'snapshot'] > ['readConcern' => new \MongoDB\Driver\ReadConcern('snapshot')]
+    * ['writeConcern' => 'majority'] > ['writeConcern' => new \MongoDB\Driver\WriteConcern('majority')]
+    * ['writeConcern' => ['majority',true]] > ['writeConcern' => new \MongoDB\Driver\WriteConcern('majority',true)]
+    * ['readPreference' => 'snapshot'] > ['readPreference' => new \MongoDB\Driver\ReadPreference('primary')]
+    */
+    public static function prepareCPOptions(&$options)
```
Comment (samdark) [inline src/Command.php:106]: Does it need to be public?

File: src/Transaction.php [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+    /**
+     * Set debug message if `enableLogging` property is enable in yii\mongodb\Connection
+     * @var string $message please see $this->yiiDebug()
+     * @var string $category please see $this->yiiDebug()
+     */
+    protected function yiiDebug($message, $category = 'mongodb')
+    {
+        if ($this->clientSession->db->enableLogging) {
+            yii::debug($message,$category);
```
Comment (samdark) [inline src/Transaction.php:50]:
```suggestion
            Yii::debug($message,$category);
```

File: src/Transaction.php [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+    /**
+     * Begin profile if `enableProfiling` property is enable in yii\mongodb\Connection
+     * @var string $token please see $this->yiiBeginProfile()
+     * @var string $category please see $this->yiiBeginProfile()
+     */
+    protected function yiiBeginProfile($token, $category = 'mongodb')
+    {
+        if ($this->clientSession->db->enableProfiling) {
+            yii::beginProfile($token,$category);
```
Comment (samdark) [inline src/Transaction.php:62]:
```suggestion
            Yii::beginProfile($token,$category);
```

File: src/Transaction.php [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+    /**
+     * End profile if `enableProfiling` property is enable in yii\mongodb\Connection
+     * @var string $token please see $this->yiiEndProfile()
+     * @var string $category please see $this->yiiEndProfile()
+     */
+    protected function yiiEndProfile($token, $category = 'mongodb')
+    {
+        if ($this->clientSession->db->enableProfiling) {
+            yii::endProfile($token,$category);
```
Comment (samdark) [inline src/Transaction.php:74]:
```suggestion
            Yii::endProfile($token,$category);
```

**Comment by samdark** [comment]
Overall it looks alright but, it seems, public interfaces are changed so that should be a major release. Right?

File: src/ActiveRecord.php [hunk 19]
```diff
@@ -243,6 +272,76 @@ protected function insertInternal($attributes = null)
...
+     * $customer->email = $email;
+     * $customer->update();
+     * ```
+     *
+     * Note that it is possible the update does not affect any row in the table.
+     * In this case, this method will return 0. For this reason, you should use the following
+     * code to check if update() is successful or not:
+     *
+     * ```php
+     * if ($customer->update() !== false) {
```
[the remark this answers is not in the document]
Comment (ziaratban) [reply src/ActiveRecord.php:309]: I copied this comment from [main source of yii2](https://github.com/yiisoft/yii2/blob/827db6c90ea3fb623a0c480c6a4c8ae46c655cf8/framework/db/ActiveRecord.php#L672).
I can not found any `null` in example. Where exactly is it?

[the remark this answers is not in the document]
Comment (samdark) [reply src/ActiveRecord.php:309]: I've meant `false`, not `null` :)

[the remark this answers is not in the document]
Comment (ziaratban) [reply src/ActiveRecord.php:309]: `@return int|false the number of rows affected, or false if validation fails`

I think everything is ok.

File: src/Command.php (hunk 15 above)
Comment (ziaratban) [reply src/Command.php:106]: Yes , used in https://github.com/yiisoft/yii2-mongodb/blob/f42ff05e7b911b6551673a35ee8bf25e3c58794e/src/Transaction.php#L106

**Comment by ziaratban** [comment]
> Overall it looks alright but, it seems, public interfaces are changed so that should be a major release. Right?

Of course!

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Merged. Thank you for working on it.

**Pull request merged**